### DISCUSSION TITLE
Migrating from psycopg2 - logging query and catching errors

### DISCUSSION THREAD

**Discussion by j-adamczyk** [psycopg/psycopg#330] [Q&A]
I have a wrapper around DB connection for `psycopg2` to a) log query b) catch and log errors (for interacting with AWS Lambda functions, but this does not matter much here). The most important method is:
```python
class DbConnector:
    def _execute(self, db_name, connection, query_dict):
        try:
            cursor = connection.cursor(cursor_factory=LoggingCursor)
            cursor.execute(query_dict["queries"], query_dict["data"])
        except psycopg2.Error as e:
            self.logger.error("[DbConnector] An error occurred, rollback transaction!")
        
            lambda_error = LambdaError(
                status=400,
                error=e.pgerror if e.pgerror else e,
                data=query_dict["queries"] % query_dict["data"],
            )
            self.logger.error(f"[DbConnector] {lambda_error}")

            connection.rollback()

            raise lambda_error

        return cursor
```

I need to migrate this to `psycopg3`. A couple questions:
1. I tried to migrate `LoggingCursor`, which basically mogrifies the query and prints it. Is this correct?
```python
class LoggingCursor(psycopg.Cursor):
    def execute(
        self,
        query: Query,
        params: Optional[Params] = None,
        *,
        prepare: Optional[bool] = None,
        binary: Optional[bool] = None,
    ):
        logger = logging.getLogger("sql_debug")

        query_string = self.mogrify(query, params)
        logger.info(query_string)

        try:
            cursor = psycopg.Cursor
            cursor.execute(self, query, params, prepare=prepare, binary=binary)
        except Exception as exc:
            logger.error(f"{exc.__class__.__name__}: {exc}")
            raise

    def mogrify(self, query, params) -> str:
        # https://github.com/psycopg/psycopg/discussions/153
        # TODO: replace with psycopg3 .mogrify() once it is released
        args_str = [psycopg.sql.Literal(p).as_string(self.connection) for p in params]
        query_string = query % tuple(args_str)
        query_string = _RE_COMBINE_WHITESPACE.sub(" ", query_string).strip()
        return query_string
```
2. I guess that `cursor = connection.cursor(cursor_factory=LoggingCursor)` has to be changed. However, for `cursor = LoggingCursor(connection)` I don't have `RowFactory` and I don't know how to implement that.
3. `except psycopg2.Error as e:` - what should I catch for `psycopg3`? Just `psycopg.Error`?
4. `e.pgerror if e.pgerror else e` - here I want the error string, basically. Is this `str(e.diag)` in `psycopg3`?

**Comment by dvarrazzo** [comment]
> 1. I tried to migrate LoggingCursor, which basically mogrifies the query and prints it. Is this correct?

This is likely wrong:
```python
        try:
            cursor = psycopg.Cursor
            cursor.execute(self, query, params, prepare=prepare, binary=binary)
        except Exception as exc:
            ...
```

if this is a cursor subclass, you want the likes of:
```python
try:
    return super().execute(query, params, prepare=prepare, binary=binary)
except Exception as exc:
    ...
```

> 2.  guess that cursor = connection.cursor(cursor_factory=LoggingCursor) has to be changed.

No: you can just set your connection's `cursor_factory` 

```python
conn = psycopg.connect(..., cursor_factory=LoggingCursor)
```

or change the [`cursor_factory`](https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.cursor_factory) attribute. Either will take care of both `connection.cursor()` and `connection.execute()`. You shouldn't need to subclass the connection.

> 3. except psycopg2.Error as e: - what should I catch for psycopg3? Just psycopg.Error?

Yes, unless you want to catch also Python errors.

> 4. e.pgerror if e.pgerror else e - here I want the error string, basically. Is this str(e.diag) in psycopg3?

`str(e)` should work. If you want only the primary line you can use [`e.diag.message_primary`](https://www.psycopg.org/psycopg3/docs/api/errors.html#psycopg.errors.Diagnostic).
